connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2724,7 +2724,7 @@ class Dataset[T] private[sql] (
   }
 
   def sameSemantics(other: Dataset[T]): Boolean = {
```

Comment:
Hm, shall we add it into Python API too if we should?

I remember this wasn't added for some concerns from @hvanhovell (maybe I am remembering this wrongly?). This is important API for ML to use in any event. cc @WeichenXu123 FYI

Reply:
In this PR I have added the python version. Can you take a look?

Reply:
It seems @hvanhovell and  @cloud-fan  had some concerns in `sameSemantics` and `semanticHash`

https://github.com/apache/spark/pull/38742#discussion_r1029446204

https://github.com/apache/spark/pull/38742#discussion_r1031014545

Reply:
I think the hash argument still stands. However I think this is also a matter of setting the right expectations here, and to update the docs accordingly. 

@WeichenXu123 it would be good to understand your usecase.